cudax/include/cuda/experimental/__async/sender/intrusive_queue.cuh

```diff
@@ -0,0 +1,319 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//                         Copyright (c) 2021-2022 Facebook, Inc & AFFILIATES.
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX_ASYNC_DETAIL_INTRUSIVE_QUEUE
+#define __CUDAX_ASYNC_DETAIL_INTRUSIVE_QUEUE
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__iterator/iterator_traits.h>
+#include <cuda/std/__utility/exchange.h>
+#include <cuda/std/__utility/swap.h>
+
+#include <cuda/experimental/__detail/config.cuh>
+
+#include <cuda/experimental/__async/sender/prologue.cuh>
+
+namespace cuda::experimental::__async
+{
+template <auto _Next>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __intrusive_queue;
+
+template <class _Item, _Item* _Item::* _Next>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __intrusive_queue<_Next>
+{
+public:
+  _CUDAX_DEFAULTED_API __intrusive_queue() noexcept = default;
+
+  _CUDAX_API __intrusive_queue(__intrusive_queue&& __other) noexcept
+      : __head_(_CUDA_VSTD::exchange(__other.__head_, nullptr))
+      , __tail_(_CUDA_VSTD::exchange(__other.__tail_, nullptr))
+  {}
+
+  _CUDAX_API __intrusive_queue(_Item* __head, _Item* __tail) noexcept
+      : __head_(__head)
+      , __tail_(__tail)
+  {}
```

Comment:
Should this assert that `__head` and `__tail` are non-null or is that a valid use case

Reply:
they could be both null, which initializes an empty intrusive queue.